Array snapshot encoding and specs

dario23 opened the pull request.

dario23 commented on prusti-viper/src/encoder/snapshot/encoder.rs:
```diff
@@ -537,6 +557,120 @@ impl SnapshotEncoder {
+                        ],
+                    );
+
+                    let seq_lookup: vir::Expr = todo!("there is no vir::Expr to express `data[idx]` viper-Seq lookup :-(");
```
this is currently my problem. i want to say
```
forall data: Seq<Snap$...>, index: Int :: {read(cons(data), index)}
      read(cons(data), index) == data[index]
}
```
but as that forall body has to be a `vir::Expr`, i can't really express `data[index]` (i.e. indexing on a viper `Seq`). do i just add `Seq` and `SeqIndex` to the `vir::Expr` enum? don't think that's the proper solution, given how long we spent discussing whether to add `ArrayAccess` to it. @fpoli @Aurel300

fpoli replied:
PR #181 defines a `Expr::ContainerOp` and a `ContainerOpKind` with various operations, but not indexing on a Viper `Seq`. To me the organization of that PR looks like a good plan that you can follow: you could add a `ContainerOpKind` with just a `SeqIndex` operation (but don't spent time adding unneeded operations). I think @Aurel300 is more familiar than me with that work. `ArrayAccess` is more a Rust concept, that's why we were reluctant with adding it to VIR.

Aurel300 replied:
I agree. VIR is both a subset and a superset of Viper ASTs. It is a subset because we haven't had the need to use all of Viper's AST yet. Adding variants which map pre-existing Viper nodes is fine. On the other hand, it is a superset because we sometimes want to express Rust/Prusti specifics. Here the decision to add new things is not always obvious, as we don't want to add ad hoc variants to solve tiny problems if they could be solved with the existing variants and better supporting code.

dario23 committed:
Add Seq, ContainerOp to vir::Expr

dario23 committed:
Implement array snapshotting

dario23 committed:
Use array snapshot encoding in pure_function_encoder

dario23 committed:
Add simple array snapshot testcase

dario23 committed:
Handle preserving array contents in loops

They can't be as easily copied as scalar values, so we use snapshots
instead.

dario23 committed:
Do not take val_fields of arrays

dario23 committed:
Add testcase for passing array to pure function

dario23 committed:
Use lookup_pure for specs

actual pure functions will have their arrays be snapshots, so we do
read(snap, idx) there, but interacting with what the surrounding
procedure knows in specification code is easier when just encoding as
lookup_pure(a, idx).

dario23 committed:
Rework labeling for array equalities in loops

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -4075,9 +4078,14 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+        let old_label = match self .array_loop_old_label.entry(loop_head) {
+            Occupied(lbl) => lbl.into_mut(),
+            Vacant(v) => v.insert(self.cfg_method.get_fresh_label_name()),
+        };
```
Is this not just:
```suggestion
        let old_label = self.array_loop_old_label.entry(loop_head)
            .or_insert_with(|| self.cfg_method.get_fresh_label_name());
```
?

dario23 replied:
that's what i first had as well, but it moves self into the closure :-/

Aurel300 replied:
Oh, that… Oh well, we can leave this. Nitpick:
```suggestion
        use std::collections::hash_map::Entry::*;
        let old_label = match self.array_loop_old_label.entry(loop_head) {
            Occupied(lbl) => lbl.into_mut(),
            Vacant(v) => v.insert(self.cfg_method.get_fresh_label_name()),
        };
```

dario23 committed:
Remove superfluous space

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

dario23 committed:
Move elem snapshot type out of EncodedArrayTypes

dario23 committed:
Don't store element snapshot type for slices either

dario23 committed:
Don't create false snapshot types in snapshot array encoding

dario23 committed:
Recursively format sequence type when computing function identifier

Aurel300 commented on prusti-common/src/vir/ast/expr_transformers.rs:
```diff
@@ -256,6 +260,26 @@ pub trait ExprFolder: Sized {
+    }
+
+    fn fold_seq(&mut self, t: Type, elems: Vec<Expr>, p: Position) -> Expr {
+        Expr::Seq(t, elems.into_iter().map(|e| self.fold(e)).collect(), p)
```
I forget, does this work in Rust?
```suggestion
        Expr::Seq(t, elems.into_iter().map(self.fold).collect(), p)
```

dario23 replied:
it seems it doesn't :-(

fpoli replied:
Probably because `self.fold(e)` has two parameters: `Self::fold(self, e)`

Aurel300 commented on prusti-common/src/vir/ast/expr_transformers.rs:
```diff
@@ -256,6 +260,26 @@ pub trait ExprFolder: Sized {
+    fn fold_seq(&mut self, t: Type, elems: Vec<Expr>, p: Position) -> Expr {
+        Expr::Seq(t, elems.into_iter().map(|e| self.fold(e)).collect(), p)
+    }
+
```
```suggestion
```

dario23 committed:
Remove unnecessary space

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

Aurel300 commented on prusti-common/src/vir/ast/expr_transformers.rs:
```diff
@@ -695,6 +740,35 @@ pub trait FallibleExprFolder: Sized {
+            folded.push(self.fallible_fold(e)?);
+        }
+
+        Ok(Expr::Seq(ty, folded, p))
```
```suggestion
        Ok(Expr::Seq(
            ty,
            elems.into_iter()
                .map(|e| self.fallible_fold(e))
                .collect::<Result<_>>()?,
            p,
        ))
```

dario23 replied:
oh, neat, didn't know collect had `Control.Monad.sequence` built-in :D

Aurel300 commented on prusti-common/src/vir/optimizations/methods/purifier.rs:
```diff
@@ -422,6 +422,7 @@ impl ast::StmtFolder for VarPurifier {
                 ast::Type::TypedRef(_) => "builtin$havoc_ref",
                 ast::Type::Domain(_) => unreachable!(),
                 ast::Type::Snapshot(_) => unreachable!(),
+                ast::Type::Seq(_) => unreachable!(),
```
```suggestion
                ast::Type::Domain(_)
                | ast::Type::Snapshot(_)
                | ast::Type::Seq(_) => unreachable!(),
```

Aurel300 commented on prusti-viper/src/encoder/foldunfold/footprint.rs:
```diff
@@ -68,6 +68,27 @@ impl ExprFootprintGetter for vir::Expr {
+                res
+            }
+
+
```
```suggestion
```

Aurel300 commented on prusti-viper/src/encoder/snapshot/encoder.rs:
```diff
@@ -388,12 +401,43 @@ impl SnapshotEncoder {
+
+        Ok(read_func.apply(
+            vec![
+                vir::Expr::snap_app(array),
```
Given we are in snapshot encoder, can we eagerly expand this `snap_app`? Or is this one of those things that will cause a double borrow?

dario23 replied:
hm, could work if we called `self.patch_snapshots_expr` directly instead of `encoder.patch_snapshots`. i'll test and report back.

dario23 replied:
maybe `self.snap_app` rather

dario23 replied:
hm.. doesn't seem to work, but not because double borrows. getting problems with (i think) double applying snapshots(?)
```
error: [Prusti internal error] consistency error: Consistency error: Function snap$__$TY$__Array$3$u32$Snap$Array$3$u32 with formal arguments List(self: Ref) cannot be applied to provided arguments List(cons$Snap$Array$3$u32$__$TY$__Seq$$int$$Snap$Array$3$u32(Seq(0, 1, 2))). (@0.0)
```

FWIW my testcase is `PRUSTI_LOG=prusti_viper::encoder::pure_function_encoder=trace,prusti_viper::encoder::snapshot::encoder=trace ./x.py run --bin prusti-rustc -- --edition 2018 -Pdump_viper_program=true prusti-tests/tests/verify/pass/arrays/array-snap.rs` if you want to have a look yourself.

i think that "duplicate" thing is becase we have a `cons :: Seq[Int] -> Snapshot(Array$...)` and `snap :: TypedRef(Array$...) -> Snapshot(Array$...)`, and even though both declare `Snapshot(Array$...)` as their return type, doing `snap(cons(..))` is a problem.

now that i think about it, we may want to get rid of one of those two, because we're doing the "pure function encoder called from procedure encoder for specifications" case with just `lookup_pure(a, i)` instead anyway, so what's currently called `cons` could just be `snap` for arrays.

Aurel300 replied:
Hm. `snap_app` (both the `Expr` variant and the `fn` in snapshot encoder) shouldn't do anything when given an already-snapshot thing. What is the input going into `encode_array_idx` here?

dario23 replied:
```
array=Local(_1: Ref(Array$3$u32), Position { line: 0, column: 0, id: 0 }), array.ty=TypedRef("Array$3$u32")
```

dario23 replied:
as the pure function encoder works backwards, it doesn't get the snapshoted thing yet, that happens only later by substituting in the state.

Aurel300 replied:
Maybe just omit the `snap_app` and see if the patcher fixes this?

dario23 replied:
seems to work, anyway.

Aurel300 commented on prusti-viper/src/encoder/snapshot/encoder.rs:
```diff
@@ -586,6 +630,162 @@ impl SnapshotEncoder {
 
+            ty::TyKind::Array(elem_ty, ..) => {
+                let elem_snap_ty = self.encode_type(encoder, elem_ty)?;
+                let at = encoder.encode_array_types(ty)?;
```
"at" again.

dario23 committed:
Expr::Seq fallible_fold without temporary variables

it's a fold, after all :-)

Aurel300 commented on prusti-viper/src/encoder/snapshot/encoder.rs:
```diff
@@ -586,6 +630,162 @@ impl SnapshotEncoder {
+                    let rhs_call = Expr::domain_func_app(
+                        cons.clone(),
+                        vec![rhs_arg.clone().into()],
+                    );
```
IIRC I also added `.apply` to domain functions? (Also some cases further down.)

dario23 replied:
ah, yes.. i even used it (two reviews above) :laughing:

dario23 committed:
Use longer `array_types` variable name consistently

dario23 committed:
Use DomainFunc::apply instead of manual Expr::domain_func_app

dario23 committed:
Simplify match

dario23 committed:
Don't explicitly snap arrays when indexing

Aurel300 commented on prusti-viper/src/encoder/foldunfold/footprint.rs:
```diff
@@ -68,6 +68,26 @@ impl ExprFootprintGetter for vir::Expr {
+                    .map(|e| {
+                        e.get_footprint(predicates)
+                    })
+                    .collect::<Vec<_>>();
```
I see there is [`FromIterator` on `HashSet`](https://doc.rust-lang.org/std/iter/trait.FromIterator.html#impl-FromIterator%3CT%3E-7) – does it work to just `.collect:<HashSet<_>>()` here?

dario23 replied:
there is, but for `T`, where `e.get_footprint` returns a `Vec<T>`. i guess i could get rid of the intermediate `elem_perms` vec, but i haven't found a simple way to merge all `HashSet`s from the iter without the temporary variables. 
i guess i could use `fold`, what do you think abou
```
            vir::Expr::Seq(_, elems, _) => {
                elems
                    .iter()
                    .map(|e| {
                        e.get_footprint(predicates)
                    })
                    .fold(HashSet::new(), |hs, fp| {
                        hs.extend(fp); hs
                    })
            }
```

Aurel300 replied:
I see, that's unfortunate. The current version works then.

Aurel300 commented on prusti-viper/src/encoder/pure_function_encoder.rs:
```diff
@@ -495,8 +495,63 @@ impl<'p, 'v: 'p, 'tcx: 'v> PureFunctionBackwardInterpreter<'p, 'v, 'tcx> {
+                    )?
+                }
+            }
+            PlaceEncoding::SliceAccess { .. } => todo!("slice access"),
```
Use an unsupported encoding error here.

Aurel300 commented on prusti-viper/src/encoder/pure_function_encoder.rs:
```diff
@@ -495,8 +495,63 @@ impl<'p, 'v: 'p, 'tcx: 'v> PureFunctionBackwardInterpreter<'p, 'v, 'tcx> {
+    fn postprocess_place_encoding(
+        &self,
+        place_encoding: PlaceEncoding<'tcx>,
+        ) -> EncodingResult<vir::Expr> {
```
```suggestion
    ) -> EncodingResult<vir::Expr> {
```

Aurel300 commented on prusti-viper/src/encoder/pure_function_encoder.rs:
```diff
@@ -1259,6 +1345,17 @@ impl<'p, 'v: 'p, 'tcx: 'v> BackwardMirInterpreter<'tcx>
+                                let array_types = self.encoder.encode_array_types(place_ty).with_span(span)?;
+                                state.substitute_value(&opt_lhs_value_place.unwrap(), array_types.array_len.into());
+                            }
+                            _ => unimplemented!("len of '{:?}'", place_ty),
```
Can we do a proper error here as well?

dario23 committed:
Shorten footprint collection for vir::Expr::Seq

dario23 committed:
Emit EncodingError::unsupported instead of panicking

dario23 committed:
Fix indent

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

dario23 committed:
Add equality to lookup_pure as array snap postcondition

dario23 committed:
Also patch snapshots in DomainFuncApp

pretty similar to how FuncApp is handled

dario23 committed:
Arrays support equality as well

dario23 committed:
Use snapshots for everything in pure function encoder

dario23 committed:
add two more testcases

dario23 committed:
Assume that snapshot read is the same as lookup

For big arrays prusti/viper had problems verifying, so we'll just go
with assuming it.
